### PR TITLE
Timesheet lockdown with grace period

### PR DESCRIPTION
**Pull request by HonzaKopecky**
## Description
Introducing feature that enables admin to setup a lockdown period after which timesheet entries are no longer editable/creatable. Using PHP relative DateTime syntax a lockdown period is defined using `lockdown_period_start` (e.g. `first day of last month 00:00:00`) and `lockdown_period_end` (e.g. `last day of last month 23:59:59`) settings.  All entries before the start are not editable. All entries inside of the period are only editable until the 'now' date is inside of the period.

Feature also introduces setting `lockdown_grace_period`. This is also PHP relative DateTime setting. It allows users to modify/create timesheet entries even after the lockdown period has ended. It doesn't allow users to edit/create entries before the lockdown period start.

Feature also introduces two new permissions. `lockdown_complete_override_timesheet` is by default assigned to admin users and it allows them to edit/create any entries no matter the lockdown/grace period. `lockdown_grace_override_timesheet` is by default assigned to admin users and allows users to edit/create entries of the last lockdown period (between start and end) even though the lockdown and grace period passed.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [x] I verified that my code applies to the guidelines (`composer kimai:code-check`)
- [ ] I updated the documentation (see [here](https://github.com/kimai/www.kimai.org/tree/master/_documentation))
- [x] I agree that this code is used in Kimai and will be published under the [MIT license](https://github.com/kevinpapst/kimai2/blob/master/LICENSE)

## Possible todos/future work
- The fact that all settings fields are required forced me to do this hacky way with space. When the setting contains just space, the getters will return null.
- The settings should probably be validated at the time of input. Can that be achieved by creating another Type (e.g. RelativeDateTimeType)? 
- Also, `lockdown_grace_period` is dependent on `lockdown_period_end`. How would one validate this at the time of input?
- Timezones. It seems to work with timesheet entries that have other timezones but it's not really clear what timezone are the settings (start, end, grace) in thus it's a bit confusing.

### CODE REVIEW

**Comment by kevinpapst** [comment]
Hi! Thanks for the PR 👍  Sounds awesome.
I have a broken internet connection @ home and cannot test the next days.
Will get back ASAP

**Comment by kevinpapst** [comment]
The command `composer kimai:pre-commit` is meant to run as pre commit hook and only checks the stuff that works rather fast. 
Integration tests take too long for a pre commit hook.

So you need to run `composer kimai:code-check` to check everything locally.

**Comment by HeinzWuert** [comment]
Reads interesting.
Will it be possible to
- create a permanent rule which
- on a specific day of the current month
- complete the entire previous month (marked as exported)?
We always have the final invoice for the previous month on the 10th of the current month. From then on, changes to the bookings are no longer permitted.
That would be great as an automatism.

**Comment by kevinpapst** [comment]
fixes #245
fixes #1605

**Comment by hmr-it-jr** [comment]
@HonzaKopecky will you finish the development on this PR?

**Comment by kevinpapst** [comment]
@HonzaKopecky that took quite a while, but I am happy with the PR now. I changed a couple of things:
- extracted timesheet validators into single classes (not directly related to this change, but it was a good moment to do it)
- added tests for validators
- renamed the new permissions
- added a constraint to validate if the configuration are valid date time strings 
- validate that end is not before start
- added documentation (see https://github.com/kimai/www.kimai.org/pull/67)

Do you want to give it a try?

File: src/Validator/Constraints/TimesheetLockdownValidator.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        // lockdown never takes effect for users with special permission
+        if ($this->auth->isGranted('lockdown_override_timesheet')) {
+            return;
+        }
+
+        if (!empty($constraint->now)) {
+            if ($constraint->now instanceof \DateTime) {
+                $now = $constraint->now;
+            } else {
```
Comment (kevinpapst) [inline src/Validator/Constraints/TimesheetLockdownValidator.php:98]: elseif (is_string())

File: src/Validator/Constraints/TimesheetOverlapping.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+final class TimesheetOverlapping extends TimesheetConstraint
+{
+    public const RECORD_OVERLAPPING = 'kimai-timesheet-overlapping-01';
+
+    protected static $errorNames = [
+        self::RECORD_OVERLAPPING => 'You already have an entry for this time.',
+    ];
+
+    public $message = 'Overlapping entries found.';
```
Comment (kevinpapst) [inline src/Validator/Constraints/TimesheetOverlapping.php:20]: unify message

**Comment by kevinpapst** [comment]
thanks @HonzaKopecky for the idea and input 👍

**Pull request merged**